### PR TITLE
Ethernet support for STM32F429

### PR DESCRIPTION
**Pull request by Ioan-Cristian**
### Pull Request Overview

This pull request adds basic Ethernet support for STM32F429. At the moment, the
peripheral supports only raw Ethernet frames and there is no support for
address filtering, hardware checksum control or chaining descriptors for Ethernet DMA.

### Testing Strategy

The Ethernet implementation was tested through basic unit tests and manually running a web server. (lwIP)

### TODO or Help Wanted

Issues:

+ In rare situations, the peripheral (or the capsule?) hangs and no frame is
processed anymore.
+ When compiling with level 2 or 3 optimizations, the web server doesn't work.

TODO:

+ Try to successfully compile and run the web server with optimizations (high priority)
+ Full UDP stack in the kernel
+ Offload workload from software to hardware
+ Chaining descriptors for Ethernet DMA.
+ Address filtering
+ Packets statistics: drop frames, bad frames, received/transmitted frames etc
for profiling

### Documentation Updated

- [x] Updated the relevant files in `/docs`

### Formatting

- [x] Ran `make prepush`.

### Updates

+ 29/06/2023: This pull request was converted to a draft, since the clock implementation was split up in #3528 
+ 28/09/2023: The original pull request #3523  was closed and replaced by this one

### CODE REVIEW

**Comment by Ioan-Cristian** [comment]
@lschuermann Sorry for closing the original pull request. This pull request replaces the original one.

I leave here the references to unresolved conversations:

https://github.com/tock/tock/pull/3523#discussion_r1300863902

**Review by bradjc — APPROVED** [review]

File: boards/nucleo_f429zi/src/main.rs [hunk 1]
```diff
@@ -44,7 +48,7 @@ const FAULT_RESPONSE: kernel::process::PanicFaultPolicy = kernel::process::Panic
 /// Dummy buffer that causes the linker to reserve enough space for the stack.
 #[no_mangle]
 #[link_section = ".stack_buffer"]
-pub static mut STACK_MEMORY: [u8; 0x2000] = [0; 0x2000];
+pub static mut STACK_MEMORY: [u8; 0x4000] = [0; 0x4000];
```
Comment (lschuermann) [inline boards/nucleo_f429zi/src/main.rs:51]: Note to future self: we'll work on bringing these stack sizes down when merging `tock-ethernet` into master, by fixing the resource (stack) consumption in the TAP ethernet driver.

**Review by lschuermann — APPROVED** [review]
This is a great quality driver, has been brewing for a long time, and works on the real HW. Let's get this into `tock-ethernet`, fix the TAP driver and then get initial networking support in master.

**Comment by lschuermann** [comment]
The CI failures on this seem to be a fluke, merging should work and hopefully pass CI. :crossed_fingers:

**Pull request merged**